### PR TITLE
provider/kubernetes: Add `NAMESPACE` to Location types.

### PR DESCRIPTION
**Pull request by lwander**
This is needed to correctly do dynamic lookups of kubernetes server groups for use by pipeline stages.

@duftler

### CODE REVIEW

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy [hunk 1]
```diff
@@ -58,6 +58,9 @@ class TargetServerGroup {
     } else if (loc.type == Location.Type.ZONE) {
       op.zone = loc.value
       op.zones = [loc.value]
+    } else if (loc.type == Location.Type.NAMESPACE) {
+      op.namespace = loc.value
```
Comment (duftler) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy]: Would really like to cleanup this pattern where we include the singular and plural of each location. Probably not the right time now though (as part of this PR, that is).

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy [hunk 2]
```diff
@@ -68,9 +71,15 @@ class TargetServerGroup {
...
       if (serverGroup.type == "gce") {
-        return new Location(type: Location.Type.ZONE, value: serverGroup.zones[0])
+        return Location.zone(serverGroup.zones[0])
+      }
+      if (serverGroup.namespace) {
+        return Location.namespace(serverGroup.namespace)
+      } else if (serverGroup.region) {
+        return Location.region(serverGroup.region)
+      } else {
+        throw new IllegalArgumentException("No known location type provided.")
```
Comment (duftler) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy]: Could probably add more detail into this exception message (at a minimum, the serverGroup should be referenced).

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy (hunk 1 above)
Comment (lwander) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy]: I'd also like to clean up the whole `location` thing. Have `location` be a field for every clouddriver operation, so we don't have to transform it constantly.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy [hunk 3]
```diff
@@ -86,9 +95,15 @@ class TargetServerGroup {
         if (!zones) {
           throw new IllegalStateException("Cannot find GCE zones in stage data ${stageData}")
         }
-        return new Location(type: Location.Type.ZONE, value: zones[0])
+        return Location.zone(zones[0])
+      }
+      if (stageData.namespace) {
```
Comment (duftler) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy]: Now that this routine is getting more complicated, can we consolidate the logic in `locationFromServerGroup()` and `locationFromStageData()`?

Comment (lwander) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy]: SGTM

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroupResolver.groovy [hunk 4]
```diff
@@ -114,12 +114,14 @@ class TargetServerGroupResolver {
 
     def tsg = tsgs.find {
       (stage.context.regions && stage.context.regions.contains(it.getLocation().value)) ||
-        (stage.context.zones && stage.context.zones.contains(it.getLocation().value))
+        (stage.context.zones && stage.context.zones.contains(it.getLocation().value)) ||
```
Comment (duftler) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroupResolver.groovy]: I think you can use:
def location = it.getLocation().value
return (stage.context.regions?.contains(location) || (stage.context.zones?.contains(location) || (stage.context.namespaces?.contains(location))

**Comment by duftler** [comment]
LGTM after addressing comments.

**Pull request merged**